datajunction-server/datajunction_server/sql/decompose.py

```diff
@@ -0,0 +1,146 @@
+"""Used for extracting measures form metric definitions."""
+from pydantic import BaseModel
+
+from datajunction_server.enum import StrEnum
+from datajunction_server.sql import functions as dj_functions
+from datajunction_server.sql.parsing.backends.antlr4 import ast, parse
+
+
+class Aggregability(StrEnum):
+    """
+    Type of allowed aggregation for a given measure.
+    """
+
+    FULL = "full"
+    LIMITED = "limited"
+    NONE = "none"
```

Comment:
What's an example of "limited aggregability"? `COUNT DISTINCT` is categorized as "limited" below, but I can't think of a way to aggregate `COUNT DISTINCT`

Reply:
Let's consider a metric definition like `COUNT(DISTINCT user_id)`. This can be decomposed into a single measure with expression `DISTINCT user_id` that can be aggregatable if the measure is calculated at least at the `user_id` level, but not if it's calculated at some other level (e.g., `country`).

Technically, if the aggregability type is `LIMITED`, the `level` should be specified to highlight the level at which the measure needs to be aggregated to in order to support the specified aggregation function. I haven't added this yet because I'm not sure how useful this is in practice.

So in the particular example of `COUNT(DISTINCT user_id)`, it can be decomposed into:
```
    - name: users_count_distinct
      expression: DISTINCT user_id
      aggregation: COUNT
      rule:
        type: LIMITED
        level: ["user_id"]
```

Reply:
Not saying this always makes sense, but I can see one doing some additional aggs (say AVG()) on top of count_distinct results over some particular dimensions. And if your data is MECE over some dimensions you could even do SUM() on top of it.

Reply:
Yeah, agreed, although it does also get super complicated to manage or indicate through metadata.

Reply:
> This can be decomposed into a single measure with expression DISTINCT user_id that can be aggregatable if the measure is calculated at least at the user_id level, but not if it's calculated at some other level (e.g., country).

OK, I see the case for "limited aggregability" now. 
But I think the example above is backward. Let's say we want to COUNT DISTINCT user_id globally. It _is_ possible to first `COUNT DISTINCT user_id GROUP BY country`, then `SUM(count_distinct_user_id)`. Essentially, GROUP BY any grain that is as or more coarse than `user_id` is allowed.

Is that right?

Reply:
> Essentially, GROUP BY any grain that is as or more coarse than user_id is allowed.

Only if the `grain` is `1:many` with the `user_id`, otherwise you may be double counting.

Reply:
@anhqle if you did `COUNT DISTINCT user_id GROUP BY country` first, then you would lose the ability to distinguish between overlapping user_ids across countries, and end up with double-counted results.

Reply:
Yeah, my unstated assumption is that user:country is many:one.

In any case, it's complicated to track that relationship as we discussed above. So in practice, "limited aggregability" only works when the measure is aggregated to _exactly_ `level`.